### PR TITLE
Fixes rebase and cherry-pick conflict resolution

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

The new merge conflict resolution wasn't compatible with rebases and cherry-picks.

**How did you fix it?**

Added a fallback on `git rev-list REBASE_HEAD/CHERRY_PICK_HEAD HEAD`, along with tests.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-core/sources/MessageName.ts [hunk 1]
```diff
@@ -95,7 +95,6 @@ export enum MessageName {
   NETWORK_UNSAFE_HTTP = 81,
   RESOLUTION_FAILED = 82,
   AUTOMERGE_GIT_ERROR = 83,
-  AUTOMERGE_LOCKFILE_VERSION_MISMATCH = 84,
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/MessageName.ts:98]: We can't remove these.

File: packages/gatsby/content/advanced/error-codes.md [hunk 2]
```diff
@@ -404,3 +404,12 @@ Yarn failed to locate a package version that could satisfy the requested range.
 - The registry may not have been set properly (so Yarn is querying the public npm registry instead of your internal one)
 
 - The version may have been unpublished (although this shouldn't be possible for the public registry)
+
+## YN0083 - `AUTOMERGE_GIT_ERROR`
+
+When autofixing merge conflicts, Yarn needs to know what are the two lockfile versions it must merge together. To do that, it'll run `git rev-parse MERGE_HEAD HEAD` and/or `git rev-parse REBASE_HEAD HEAD`, depending on the situation. If both of those commands fail, the merge cannot succeed.
```
Comment (merceyz) [inline packages/gatsby/content/advanced/error-codes.md:410]: There is also `CHERRY_PICK_HEAD` for `git cherry-pick` conflicts, or do you plan on doing that in another PR?

**Review by merceyz — CHANGES_REQUESTED** [review]

File: packages/yarnpkg-core/sources/MessageName.ts (hunk 1 above)
Comment (arcanis) [reply packages/yarnpkg-core/sources/MessageName.ts:98]: Yep, that's why I removed it, it's a leftover from a previous implementation.

File: packages/gatsby/content/advanced/error-codes.md (hunk 2 above)
Comment (arcanis) [reply packages/gatsby/content/advanced/error-codes.md:410]: Forgot this one, will do it in the same diff.

**Review by merceyz — APPROVED** [review]

**Pull request merged**